pkgs/by-name/li/libfprint-2-tod1-broadcom/package.nix

```diff
@@ -0,0 +1,81 @@
+{
+  autoPatchelfHook,
+  fetchzip,
+  lib,
+  libfprint-tod,
+  openssl,
+  patchelfUnstable,
+  stdenv,
+}:
+
+# Based on ideas from (using a wrapper library to redirect fopen() calls to firmware files):
+#   * https://tapesoftware.net/replace-symbol/
+#   * https://github.com/NixOS/nixpkgs/pull/260715
+let
+  version = "5.12.018";
+  src = fetchzip {
+    url = "http://dell.archive.canonical.com/updates/pool/public/libf/libfprint-2-tod1-broadcom/libfprint-2-tod1-broadcom_${version}.orig.tar.gz";
+    hash = "sha256-0C2PpYpEJNrU+8NT95w4QV0J5nHQisMY94Czw3jQOzw=";
+  };
+
+  unpacked = stdenv.mkDerivation {
+    pname = "libfprint-2-tod1-broadcom-unpacked";
+    inherit src version;
+
+    dontBuild = true;
+
+    installPhase = ''
+      mkdir $out
+      cp -r . $out
+    '';
+  };
+
+  # wraps `fopen()` for finding firmware files
+  wrapperLibName = "wrapper-lib.so";
+  wrapperLib = stdenv.mkDerivation {
+    pname = "libfprint-2-tod1-broadcom-wrapper-lib";
+    inherit version;
+
+    src = ./.;
+
+    postPatch = ''
+      substitute wrapper-lib.c lib.c \
+        --subst-var-by to "${unpacked}/var/lib/fprint/fw"
+      cc -fPIC -shared lib.c -o ${wrapperLibName}
+    '';
+
+    installPhase = ''
+      install -D -t $out/lib ${wrapperLibName}
+    '';
+  };
+in stdenv.mkDerivation {
+  pname = "libfprint-2-tod1-broadcom";
+  inherit src version;
+
+  buildInputs       = [ libfprint-tod openssl wrapperLib ];
+  nativeBuildInputs = [ autoPatchelfHook patchelfUnstable ];
+
+  installPhase = ''
```

Comment:
I see your point with the hooks, thanks for pointing that out.

Concerning using the default `installPhas` in this case: This might be a stupid question, but what does the default `installPhase` actually do if there is no Makefile (as in this case; the downloaded archive contains only an `install.sh` script for installation)? I was under the impression that, without a Makefile, it does nothing (which is what the [manual seems to indicate](https://nixos.org/manual/nixpkgs/stable/#ssec-install-phase)).

Reply:
It will not run `make install` but hooks will still be called:

https://github.com/NixOS/nixpkgs/blob/e5bc1cb431307c0a3943615151aeff0b9cc246c3/pkgs/stdenv/generic/setup.sh#L1449-L1452

Reply:
Thanks for the pointer. For now I just added the hooks. But I assume what you actually suggest is to not use `installPhase` at all. Instead of "installing" the files (like the dynamic library) I can directly use those in the unpacked driver archive placed in the nix store. Will try this tomorrow or so.

Reply:
@jtojnar: After having a closer look, I'm not sure how to

> […] rely on the default `installPhase`. […]

for the main derivation as you suggested. I first thought you meant to directly rely on the path to the dynamic library from the unpacked source, since the wrapper library has a runtime dependency on the unpacked source anyway, but that seems not very clean and the main derivation still needs to install the udev rules in  `$out/lib/udev/rules.d/` for `services.udev.packages` to work correctly, right?

I also checked the other libfprint-2-tod1-* packages (e.g., [libfprint-2-tod1-vfs0090](https://github.com/NixOS/nixpkgs/blob/681e984f8e3d66f4fd8a94ea28bb7a8b87b27216/pkgs/development/libraries/libfprint-2-tod1-vfs0090/default.nix#L24-L31), which use a similar overwrite of the `installPhase` as I do now (with the proper hooks).

So I currently would simply leave the overwritten `installPhase` as is (now with the proper hooks). But let me know if I'm overlooking something.

Reply:
> […] rely on the default `installPhase`. […]

Sorry, that indeed does not apply here. It is part of my reply template and I forgot to remove the part here this time.

---

Two derivation approach is needed here since we cannot have store path references in fixed-output derivations llike `fetchzip`.

Reply:
Ah, no wonder I couldn't figure it out 😅.

Then I'm marking this as resolved. Let me know if there's anything else I should address.